org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java

```diff
@@ -0,0 +1,237 @@
+/**
+ * Copyright (c) 2022 Sebastian Thomschke and others.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.tm4e.ui.internal.utils;
+
+import static org.eclipse.tm4e.core.internal.utils.NullSafetyHelper.*;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.tm4e.core.model.ITMModel;
+import org.eclipse.tm4e.core.model.ModelTokensChangedEvent;
+import org.eclipse.tm4e.core.model.TMToken;
+import org.eclipse.tm4e.ui.TMUIPlugin;
+import org.eclipse.tm4e.ui.internal.model.TMDocumentModel;
+
+public final class MarkerUtils {
+
+	private abstract static class AbstractMarkerConfig {
+		final String type;
+		final int priority;
+		final int severity;
+
+		AbstractMarkerConfig(final String type, final int priority, final int severity) {
+			this.type = type;
+			this.priority = priority;
+			this.severity = severity;
+		}
+	}
+
+	private static final class ProblemMarkerConfig extends AbstractMarkerConfig {
```

Comment:
I don't like the idea of having subclasses just to specialize constructors. The parent could instead provide some `newProblemMarkerConfig(priority)` and `newTaskMarkerConfig(priorty)` methods.